target/arm/cpu.h

```diff
@@ -1074,11 +1126,18 @@ int fp_exception_el(CPUARMState *env, int cur_el);
 int sve_exception_el(CPUARMState *env, int cur_el);
 uint32_t sve_zcr_len_for_el(CPUARMState *env, int el);
 
+#ifdef TARGET_CHERI
+    // Morello does not support 32-bit, so might as well optimise everything away
+    #define is_a64(env) 1
```

Comment:
Having said that, Morello QEMU would be an interesting place to experiment with getting freebsd32 to work with CHERI.

Reply:
I agree, but it would probably require a significant amount of effort. Until then optimizing the checks out seems good to me.

You might also want to remove ` bool aarch64` from CPUARMState for TARGET_CHERI.

Reply:
This is one of the things that I am currently unhappy with, doing this was not quite correct. Some places in code still access aarch64 directly (and I think I have gone and fixed all of those as well). Things currently work, but it would probably be preferable to leave things as they were, and just stop the CPU going out of a64 mode (I don't know how arm AARCH64 actually does this, and I have not found anything in the morello spec as to whether trying should be an exception or just just not do anything).

Reply:
The base AArch64 spec (in D1.11.2) says this is an illegal return event, which means PSTATE.IL is set to 1, causing an Illegal Execution fault on the next instruction fetch.

(and SPSR_ELx.M[4] is the bit that determines AArch64 (0) vs AArch32 (1), FWIW)